blockfetch: rollback recovery requests a range the peer reports as not found, stalling the node in a re-sync loop

## Summary

During a series of ten back-to-back canonical DevNet runs on main at ec9cab4d, one run
failed with a blockfetch rollback-recovery signature that appeared in none of the other
nine. Two nodes asked a peer for a block range the peer reported as not found immediately
after a rollback, retried, then dropped their queued headers and forced a chainsync
re-sync. The network never reached slot 50 inside the first scenario's budget, and the
three consensus scenarios failed as a result.

All four containers stayed healthy for the whole run, so this is not a crash or a restart.

## Observed

Every warning below is from the same run, and all six land inside 17 milliseconds:

```
12:36:19.011 dingo-relay WARN blockfetch continuation failed, retrying on alternate connection
12:36:19.012 dingo-relay WARN blockfetch could not obtain queued range, dropping queued headers and requesting chainsync re-sync
12:36:19.020 dingo-2     WARN failed to start blockfetch after fork rollback
                              error="request block range: block(s) not found"
12:36:19.022 dingo-2     WARN failed to start blockfetch after local rollback recovery
                              error="request block range: block(s) not found"
12:36:19.028 dingo-2     WARN blockfetch could not obtain queued range, dropping queued headers and requesting chainsync re-sync
                              remaining_headers=1 range_start_slot=72
12:36:19.028 dingo-2     WARN failed to start blockfetch after local rollback recovery
                              error="request block range: block(s) not found"
```

Sources: `ledger/chainsync.go:2908` (failed to start blockfetch after fork rollback) and
`ledger/chainsync.go:3162` (could not obtain queued range).

The affected connection on dingo-2 is `172.20.0.14:3001<->172.20.0.13:41066`, so the peer
whose range lookup failed is dingo-1. dingo-relay hit the same thing on
`172.20.0.16:33044<->172.20.0.14:3001`.

## Resulting scenario failures

```
--- FAIL: TestChainGrowthRate         dingo-1 did not reach slot 50 within 1m2.5s
--- FAIL: TestSustainedConsensus      no node reached slot 65 within 32.5s
--- FAIL: TestEpochBoundaryConsensus  nodes did not reach consensus within 25s (tolerance: 1 slots)
```

48 pass / 3 fail. The failures are all early-slot liveness, consistent with two of the four
nodes being stuck in a blockfetch retry and re-sync loop rather than extending the chain.

## Why this is not host load

The run series gated each run on a quiet host and sampled 1-minute load every 30 seconds.
The failing run's early-window load was 8.31, 13.10, 10.21, 9.01, 10.23 on a 128-core
machine. Two runs that passed had heavier load over the same window:

| Run | Result | First five load samples | Max |
|---|---|---|---|
| 5 | fail | 8.31 13.10 10.21 9.01 10.23 | 13.10 |
| 9 | pass | 11.13 14.62 13.63 10.86 8.89 | 19.14 |
| 3 | pass | 5.00 7.03 7.62 6.86 5.67 | 32.81 |

A separate run in the same series (run 2) did fail under the heaviest early load of the
batch, 11.15 rising to 16.32, with a foreign process taking about ten cores. That one is
attributable to load and reported here only to distinguish it: it failed
TestChainGrowthRate alone with a throughput message and produced none of the blockfetch
warnings above.

## Frequency

One occurrence in ten consecutive runs on ec9cab4d, all-Dingo topology (three producers
plus a relay), stock NtC ports. The other nine runs contain zero instances of
`failed to start blockfetch`, `could not obtain queued range`, or
`blockfetch continuation failed`.

## Distinct from existing reports

Not #3180, where all four nodes held an identical tip, logged nothing, and recovered on
their own. Also distinct from a separate single observation on this commit where dingo-2
and dingo-relay stalled at slot 493 while dingo-1 and dingo-3 reached 701, whose only
warning was `event delivery stalled: subscriber not draining type=ledger.blockfetch`. That
divergence did not recur anywhere in these ten runs.

#2248 covered peers stuck on a missing point and is closed. #1960 is slow-peer blockfetch
switching, which is a different concern.

## Reproducing

```
./internal/test/devnet/run-tests.sh
```

Repeat runs; this appeared once in ten. Setting `DEVNET_ARTIFACT_DIR` preserves the
per-service logs, container status, and generated genesis on failure, which is what
captured the sequence above. The default `docker compose logs --tail=100` window is too
narrow to reach the event once a run has progressed.

## Suggested starting point

The pairing of `fork rollback` and `local rollback recovery` failing on the same
connection microseconds apart, with `remaining_headers=1` and a low `range_start_slot`,
suggests the recovery path recomputes a fetch range against a chain state the serving peer
has already moved past, rather than a genuinely absent block. Worth checking what range
`startQueuedBlockfetchLocked` derives after a rollback and whether the peer is expected to
still hold `range_start_slot` at that moment.

--- a/ledger/chainsync.go
+++ b/ledger/chainsync.go
@@ -2238,6 +2238,13 @@ func (ls *LedgerState) RecoverAfterLocalRollback(
 							"error",
 							startErr,
 						)
+						// Retarget the selection too, not just this
+						// request: nextBlockfetchConnId prefers
+						// selectedBlockfetchConnId, so leaving it on the
+						// failed recovery connection sends the next batch
+						// of a multi-batch replay straight back to it.
+						// Matches the fallback in handleEventChainsync.
+						ls.selectedBlockfetchConnId = *activeConnId
 						startErr = ls.startQueuedBlockfetchLocked(
 							*activeConnId,
 							&pending,

--- a/ledger/chainsync.go
+++ b/ledger/chainsync.go
@@ -2273,6 +2273,14 @@ func (ls *LedgerState) RecoverAfterLocalRollback(
 		}
 		return LocalRollbackRecoveryResult{Recovered: true}
 	}
+	// Every preferred connection failed. Clear the selection rather than
+	// leaving it on a connection known not to serve this range, matching what
+	// handleEventChainsync does when its own fallbacks are exhausted. Each
+	// successful replay reassigns this field before its blockfetch attempt, so
+	// nothing downstream currently reads a stale value -- but that makes the
+	// invariant depend on every future writer reassigning first, which is not a
+	// property worth relying on.
+	ls.selectedBlockfetchConnId = ouroboros.ConnectionId{}
 	return LocalRollbackRecoveryResult{}
 }
 

--- a/ledger/chainsync.go
+++ b/ledger/chainsync.go
@@ -2216,9 +2216,11 @@ func (ls *LedgerState) RecoverAfterLocalRollback(
 			"header_count", headerCount,
 		)
 		ls.chainsyncBlockfetchMutex.Lock()
+		var startErr error
 		if ls.chainsyncBlockfetchReadyChan == nil &&
 			!ls.blockfetchContinuationPending {
-			if err := ls.startQueuedBlockfetchLocked(connId, &pending); err != nil {
+			startErr = ls.startQueuedBlockfetchLocked(connId, &pending)
+			if startErr != nil {
 				// Recovery connection may have closed. Retry with
 				// the current active best peer before giving up,
 				// otherwise the pipeline stalls until restart.
@@ -2234,16 +2236,15 @@ func (ls *LedgerState) RecoverAfterLocalRollback(
 							"active_connection_id",
 							activeConnId.String(),
 							"error",
-							err,
+							startErr,
 						)
-						if retryErr := ls.startQueuedBlockfetchLocked(*activeConnId, &pending); retryErr == nil {
-							err = nil
-						} else {
-							err = retryErr
-						}
+						startErr = ls.startQueuedBlockfetchLocked(
+							*activeConnId,
+							&pending,
+						)
 					}
 				}
-				if err != nil {
+				if startErr != nil {
 					ls.config.Logger.Warn(
 						"failed to start blockfetch after local rollback recovery",
 						"component",
@@ -2251,12 +2252,18 @@ func (ls *LedgerState) RecoverAfterLocalRollback(
 						"connection_id",
 						connId.String(),
 						"error",
-						err,
+						startErr,
 					)
 				}
 			}
 		}
 		ls.chainsyncBlockfetchMutex.Unlock()
+		if startErr != nil {
+			// Do not report recovery when the replayed headers could not
+			// be fetched. The caller must close these sessions so peer
+			// governance can establish a fresh chainsync intersection.
+			continue
+		}
 		return LocalRollbackRecoveryResult{Recovered: true}
 	}
 	return LocalRollbackRecoveryResult{}

--- a/ledger/chainsync_rollback_test.go
+++ b/ledger/chainsync_rollback_test.go
@@ -1345,6 +1345,54 @@ func TestRecoverAfterLocalRollbackReplaysPeerHeaderHistory(
 	fixture.ls.blockfetchRequestRangeCleanup()
 }
 
+func TestRecoverAfterLocalRollbackReportsBlockfetchFailure(
+	t *testing.T,
+) {
+	fixture := newChainsyncRollbackFixture(t)
+	require.NoError(t, fixture.ls.chain.Rollback(fixture.ancestorTip.Point))
+	require.NoError(t, fixture.ls.db.SetTip(fixture.ancestorTip, nil))
+	fixture.ls.currentTip = fixture.ancestorTip
+	fixture.ls.currentTipBlockNonce = append(
+		[]byte(nil),
+		fixture.ancestorNonce...,
+	)
+	requestCount := 0
+	fixture.ls.config.BlockfetchRequestRangeFunc = func(
+		ouroboros.ConnectionId,
+		ocommon.Point,
+		ocommon.Point,
+	) error {
+		requestCount++
+		return errBlockfetchNoBlocks
+	}
+
+	header := mockHeader{
+		hash:        lcommon.NewBlake2b256(testHashBytes("rollback-no-blocks")),
+		prevHash:    lcommon.NewBlake2b256(fixture.ancestorTip.Point.Hash),
+		blockNumber: fixture.ancestorTip.BlockNumber + 1,
+		slot:        fixture.ancestorTip.Point.Slot + 1,
+	}
+	fixture.ls.recordPeerHeaderHistory(ChainsyncEvent{
+		ConnectionId: fixture.connId,
+		Point:        ocommon.NewPoint(header.SlotNumber(), header.Hash().Bytes()),
+		Tip: ochainsync.Tip{
+			Point:       ocommon.NewPoint(header.SlotNumber(), header.Hash().Bytes()),
+			BlockNumber: header.BlockNumber(),
+		},
+		BlockHeader: header,
+	})
+
+	result := fixture.ls.RecoverAfterLocalRollback(
+		[]ouroboros.ConnectionId{fixture.connId},
+		fixture.ancestorTip.Point,
+	)
+
+	assert.False(t, result.Recovered)
+	assert.False(t, result.SkipConnectionClose)
+	assert.Equal(t, 1, requestCount)
+	assert.Equal(t, 1, fixture.ls.chain.HeaderCount())
+}
+
 func TestRecoverAfterLocalRollbackResetsStateWithoutTrackedClients(
 	t *testing.T,
 ) {

--- a/ledger/chainsync_rollback_test.go
+++ b/ledger/chainsync_rollback_test.go
@@ -1345,6 +1345,93 @@ func TestRecoverAfterLocalRollbackReplaysPeerHeaderHistory(
 	fixture.ls.blockfetchRequestRangeCleanup()
 }
 
+// TestRecoverAfterLocalRollbackRetargetsSelectedBlockfetchConn asserts the
+// active-peer fallback moves the blockfetch selection, not just the one request
+// it issues. nextBlockfetchConnId prefers selectedBlockfetchConnId, so a
+// selection left on the failed recovery connection sends the next batch of a
+// multi-batch replay straight back to the connection that just failed, and the
+// first batch is the only one that ever arrives.
+func TestRecoverAfterLocalRollbackRetargetsSelectedBlockfetchConn(
+	t *testing.T,
+) {
+	fixture := newChainsyncRollbackFixture(t)
+	require.NoError(t, fixture.ls.chain.Rollback(fixture.ancestorTip.Point))
+	require.NoError(t, fixture.ls.db.SetTip(fixture.ancestorTip, nil))
+	fixture.ls.currentTip = fixture.ancestorTip
+	fixture.ls.currentTipBlockNonce = append(
+		[]byte(nil),
+		fixture.ancestorNonce...,
+	)
+
+	activeConnId := ouroboros.ConnectionId{
+		LocalAddr:  &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 6001},
+		RemoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 3002},
+	}
+	fixture.ls.config.GetActiveConnectionFunc = func() *ouroboros.ConnectionId {
+		return &activeConnId
+	}
+	// The recovery connection is gone; only the active best peer answers.
+	var requested []ouroboros.ConnectionId
+	fixture.ls.config.BlockfetchRequestRangeFunc = func(
+		connId ouroboros.ConnectionId,
+		_ ocommon.Point,
+		_ ocommon.Point,
+	) error {
+		requested = append(requested, connId)
+		if sameConnectionId(connId, activeConnId) {
+			return nil
+		}
+		return errBlockfetchNoBlocks
+	}
+
+	header := mockHeader{
+		hash: lcommon.NewBlake2b256(
+			testHashBytes("rollback-retarget"),
+		),
+		prevHash:    lcommon.NewBlake2b256(fixture.ancestorTip.Point.Hash),
+		blockNumber: fixture.ancestorTip.BlockNumber + 1,
+		slot:        fixture.ancestorTip.Point.Slot + 1,
+	}
+	fixture.ls.recordPeerHeaderHistory(ChainsyncEvent{
+		ConnectionId: fixture.connId,
+		Point: ocommon.NewPoint(
+			header.SlotNumber(),
+			header.Hash().Bytes(),
+		),
+		Tip: ochainsync.Tip{
+			Point: ocommon.NewPoint(
+				header.SlotNumber(),
+				header.Hash().Bytes(),
+			),
+			BlockNumber: header.BlockNumber(),
+		},
+		BlockHeader: header,
+	})
+
+	fixture.ls.RecoverAfterLocalRollback(
+		[]ouroboros.ConnectionId{fixture.connId},
+		fixture.ancestorTip.Point,
+	)
+
+	require.Len(
+		t, requested, 2,
+		"the failed recovery connection then the active best peer",
+	)
+	assert.True(
+		t, sameConnectionId(requested[1], activeConnId),
+		"the fallback request must go to the active best peer",
+	)
+	assert.True(
+		t,
+		sameConnectionId(
+			fixture.ls.selectedBlockfetchConnId,
+			activeConnId,
+		),
+		"the blockfetch selection must follow the fallback, so the next "+
+			"batch does not return to the failed connection",
+	)
+}
+
 func TestRecoverAfterLocalRollbackReportsBlockfetchFailure(
 	t *testing.T,
 ) {

--- a/ledger/chainsync_rollback_test.go
+++ b/ledger/chainsync_rollback_test.go
@@ -1432,6 +1432,77 @@ func TestRecoverAfterLocalRollbackRetargetsSelectedBlockfetchConn(
 	)
 }
 
+// TestRecoverAfterLocalRollbackClearsSelectionWhenEveryConnectionFails asserts
+// the blockfetch selection is not left pointing at a connection that just failed
+// to serve the replayed range. handleEventChainsync clears it when its own
+// fallbacks are exhausted; this path now does the same, so nothing downstream
+// has to depend on being the next writer of the field.
+func TestRecoverAfterLocalRollbackClearsSelectionWhenEveryConnectionFails(
+	t *testing.T,
+) {
+	fixture := newChainsyncRollbackFixture(t)
+	require.NoError(t, fixture.ls.chain.Rollback(fixture.ancestorTip.Point))
+	require.NoError(t, fixture.ls.db.SetTip(fixture.ancestorTip, nil))
+	fixture.ls.currentTip = fixture.ancestorTip
+	fixture.ls.currentTipBlockNonce = append(
+		[]byte(nil),
+		fixture.ancestorNonce...,
+	)
+
+	activeConnId := ouroboros.ConnectionId{
+		LocalAddr:  &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 6002},
+		RemoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 3003},
+	}
+	fixture.ls.config.GetActiveConnectionFunc = func() *ouroboros.ConnectionId {
+		return &activeConnId
+	}
+	// Nothing can serve the range: neither the recovery connection nor the
+	// active best peer the fallback reaches for.
+	fixture.ls.config.BlockfetchRequestRangeFunc = func(
+		ouroboros.ConnectionId,
+		ocommon.Point,
+		ocommon.Point,
+	) error {
+		return errBlockfetchNoBlocks
+	}
+
+	header := mockHeader{
+		hash: lcommon.NewBlake2b256(
+			testHashBytes("rollback-clear-selection"),
+		),
+		prevHash:    lcommon.NewBlake2b256(fixture.ancestorTip.Point.Hash),
+		blockNumber: fixture.ancestorTip.BlockNumber + 1,
+		slot:        fixture.ancestorTip.Point.Slot + 1,
+	}
+	fixture.ls.recordPeerHeaderHistory(ChainsyncEvent{
+		ConnectionId: fixture.connId,
+		Point: ocommon.NewPoint(
+			header.SlotNumber(),
+			header.Hash().Bytes(),
+		),
+		Tip: ochainsync.Tip{
+			Point: ocommon.NewPoint(
+				header.SlotNumber(),
+				header.Hash().Bytes(),
+			),
+			BlockNumber: header.BlockNumber(),
+		},
+		BlockHeader: header,
+	})
+
+	result := fixture.ls.RecoverAfterLocalRollback(
+		[]ouroboros.ConnectionId{fixture.connId},
+		fixture.ancestorTip.Point,
+	)
+	require.False(t, result.Recovered)
+
+	assert.Empty(
+		t, connIdKey(fixture.ls.selectedBlockfetchConnId),
+		"an exhausted recovery must not leave the selection on a "+
+			"connection that failed to serve the range",
+	)
+}
+
 func TestRecoverAfterLocalRollbackReportsBlockfetchFailure(
 	t *testing.T,
 ) {

